diff --git a/Makefile b/Makefile
index d8d9e97..5c3028f 100644
--- a/Makefile
+++ b/Makefile
@@ -32,7 +32,7 @@ OFILES := $(BL1_OFILES) $(CFILES:src/%.c=build/%.o)
 ifeq ($(DEBUG),1)
 	CFLAGS := -O0
 else
-	CFLAGS := -Os
+	CFLAGS := -O1
 endif
 
 CFLAGS += -g3 -Wall -Wno-unused-function -std=gnu99 -ffreestanding -fdata-sections -ffunction-sections -nostdinc -D__KERNEL__
diff --git a/fuse.sh b/fuse.sh
index 03b2be8..c861a9f 100755
--- a/fuse.sh
+++ b/fuse.sh
@@ -15,20 +15,13 @@
 # along with this program; if not, write to the Free Software
 # Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 
-# Automatically re-run script under sudo if not root
-if [ $(id -u) -ne 0 ]; then
-  echo "Rerunning script under sudo..."
-  sudo "$0" "$@"
-  exit
-fi
-
 if [ -z $1 ]; then
 	echo "Usage: $0 DEVICE"
 	exit 0
 fi
 
 case $1 in
-/dev/sd[a-z] | /dev/loop0)
+/dev/sd[a-z] | /dev/loop0 | /dev/mmcblk[0-9])
 	if [ ! -e $1 ]; then
 		echo "Error: $1 does not exist."
 		exit 1
@@ -68,12 +61,18 @@ let BL2_POSITION=${BL1_POSITION}-${BL2_SIZE}-${ENV_SIZE}
 # ----------------------------------------------------------
 # Create a binary for movinand/mmc boot
 
-# pad to 256k
-dd if=/dev/zero bs=1k count=256 2> /dev/null | tr "\000" "\377" > build/nanoboot-256k.bin
-dd if=build/nanoboot.bin of=build/nanoboot-256k.bin conv=notrunc 2> /dev/null
+if [ -f nanoboot.bin ]; then
+	NANOBOOT=nanoboot.bin
+else
+	NANOBOOT=/usr/share/nanopi-nanoboot/nanoboot.bin
+fi
+
+[ -f ${NANOBOOT} ] || {
+	echo 'Cannot find nanoboot.bin file'
+	exit 1
+}
 
-dd if=build/nanoboot-256k.bin of=/dev/${DEV_NAME} bs=512 seek=${BL2_POSITION} conv=fdatasync &> /dev/null
-dd if=build/nanoboot-256k.bin of=/dev/${DEV_NAME} bs=512 seek=${BL1_POSITION} count=16 conv=fdatasync &> /dev/null
-rm build/nanoboot-256k.bin
+dd if=${NANOBOOT} of=/dev/${DEV_NAME} bs=512 seek=${BL2_POSITION} conv=fdatasync status=none
+dd if=${NANOBOOT} of=/dev/${DEV_NAME} bs=512 seek=${BL1_POSITION} count=16 conv=fdatasync status=none
 
-echo "nanoboot fused"
+echo 'nanoboot fused'
